### DISCUSSION TITLE
xr.open_mfdataset raised duplicate values

### DISCUSSION THREAD

**Discussion by shuai-zhou** [pydata/xarray#6297] [Q&A] [answered]
I was trying to read multiple NetCDF files downloaded from [here](https://www.ncei.noaa.gov/thredds/catalog/cdr/ndvi/files/2016/catalog.html) use `ds = xr.open_mfdataset('file_path', decode_coords='all', decode_times=False)`. I got an error: `cannot reindex or align along dimension 'latitude' because the index has duplicate values`.

A similar discussion can be found [here](https://github.com/pydata/xarray/discussions/5895). Like the situation in the discussion, the data I was trying to read has missing values, so why do the missing value issue raise duplicate values error? @TomNicholas mentioned dropping duplicates, but `.drop_duplicates` only works with data array, so how can I call `.drop_duplicates` to `preprocess`? Thanks.

**Comment by TomNicholas** [comment, marked as answer]
You could either `.map` the `drop_duplicates` method over the variables in the dataset, or just use [the code in the `drop_duplicates` method](https://github.com/pydata/xarray/blob/de965f342e1c9c5de92ab135fbc4062e21e72453/xarray/core/dataarray.py#L4660) directly on the dataset. Then you can create your own function to use within `preprocess` like this

```python
def drop_duplicates(obj, dim, keep="first"):
    if dim not in obj.dims:
        raise ValueError(f"'{dim}' not found in dimensions")
    indexes = {dim: ~obj.get_index(dim).duplicated(keep=keep)}
    return obj.isel(indexes)
```

Given that this works on datasets as well as dataarrays I don't know why there isn't a `Dataset.drop_duplicates` method - seems like we could add one.

**Reply by shuai-zhou** [reply]
Thanks. Could you please elaborate on the process? I did the following:

```python
def drop_duplicates(obj, dim, keep="first"):
    if dim not in obj.dims:
        raise ValueError(f"'{dim}' not found in dimensions")
    indexes = {dim: ~obj.get_index(dim).duplicated(keep=keep)}
    return obj.isel(indexes)

ds = xr.open_mfdataset('file_path_*.nc',
                        decode_coords="all",
                        decode_times=False,
                        preprocess=drop_duplicates(obj, 'latitude'))
```

And I got an error `name 'obj' is not defined`, where should I define the `obj`? Thanks.

**Reply by TomNicholas** [reply]
Hi, you need to pass the function handle, so if you're always going to call it with the same dim it would be

```python
def drop_latitude_duplicates(obj, dim="latitude", keep="first"):
    if dim not in obj.dims:
        raise ValueError(f"'{dim}' not found in dimensions")
    indexes = {dim: ~obj.get_index(dim).duplicated(keep=keep)}
    return obj.isel(indexes)


ds = xr.open_mfdataset('file_path_*.nc',
                        decode_coords="all",
                        decode_times=False,
                        preprocess=drop_latitude_duplicates)
```
or if you want to be able to choose which dim later on then you need to use `functools.partial` to bind the other arguments in like this 
```python
import functools


def drop_latitude_duplicates(obj, dim, keep="first"):
    if dim not in obj.dims:
        raise ValueError(f"'{dim}' not found in dimensions")
    indexes = {dim: ~obj.get_index(dim).duplicated(keep=keep)}
    return obj.isel(indexes)


ds = xr.open_mfdataset('file_path_*.nc',
                        decode_coords="all",
                        decode_times=False,
                        preprocess=functools.partial(drop_duplicates, dim="latitude", keep="first"))
```

**Reply by shuai-zhou** [reply]
Thanks for your prompt response. Both methods work (the `def drop_latitude_duplicates` in the second method should be `def drop_duplicates`). But there is another error saying `cannot reindex or align along dimension 'longitude' because the index has duplicate values`. So I do not know what dimensions have duplicates or if other dimensions have duplicates. Could those methods drop duplicates for every dimension? Thanks.

**Reply by TomNicholas** [reply]
You could just loop over all dimensions, something like this

```python
def drop_duplicates_along_all_dims(obj, keep="first"):
    deduplicated = obj
    for dim in obj.dims:
        indexes = {dim: ~deduplicated.get_index(dim).duplicated(keep=keep)}
        deduplicated = deduplicated.isel(indexes)
    return deduplicated

ds = xr.open_mfdataset('file_path_*.nc',
                        decode_coords="all",
                        decode_times=False,
                        preprocess=drop_duplicates_along_all_dims)
```
(though perhaps we should promote `.drop_duplicates` to accept a list of dims, or `...` for all dims).

**Reply by TomNicholas** [reply]
Actually you don't even need the loop, you can just build a dict of indexes and pass to `isel` once, i.e.

```python
def drop_duplicates_along_all_dims(obj, keep="first"):
    all_dims = self.dims
    indexes = {dim: ~self.get_index(dim).duplicated(keep=keep) for dim in all_dims}
    return self.isel(indexes)
```

**Reply by TomNicholas** [reply]
(Also see #6307)

**Reply by shuai-zhou** [reply]
It works. Thanks!
